tools/rpcgen.py

```diff
@@ -103,6 +105,37 @@ class failure_probes;
       {%- endfor %}
     }
 
+    void setup_public_metrics() final {
+        namespace sm = ss::metrics;
+        auto component_label = sm::label("component");
+        auto service_label = sm::label("service");
+        auto method_label = sm::label("method");
+      {%- for method in methods %}
+        {
+            std::vector<ss::metrics::label_instance> labels{
+              service_label("{{service_name}}"),
+              method_label("{{method.name}}"),
+              component_label("internal")};
+                auto aggregate_labels
+                  = config::shard_local_cfg().aggregate_metrics()
+                      ? std::vector<sm::label>{sm::shard_label, method_label}
+                      : std::vector<sm::label>{};
+
+            _public_metrics.add_group(
+              prometheus_sanitize::metrics_name("rpc"),
+              {sm::make_histogram(
```

Comment:
Is latency measurements here necessary at the moment? I was hesitant to add them here because:
1. The metric from our specification is `redpanda_rpc_request_latency_seconds` which implies latency measurements coming from "requests" only. If I read the code base correctly, the methods from a rpc::service (such as those in this file) are **not** invoked because of an incoming request.
2. Each service has several methods and those increase the dimensionality. Aggregating on the method does help some: /public_metrics has 1024 lines w/o aggregation and 268 lines w/ aggregation.

IMO, we can remove latency from this section for now unless it is specifically requested. Thoughts?

Reply:
I don't really understand the difference between this and `redpanda_rpc_request_latency_seconds`, except that this has `method`, which we'd want to aggregate away.

I don't know if it was intended to distinguish "requests", from non-incoming requests? Are we talking about things like heartbeats? In any case, I think it's still an rpc, which is a request. Maybe request is duplicitous when rpc already exists in the name.

Reply:
Going forward, going to leave this in and aggregate on service. Hopefully no double registration problem.

Reply:
So, is this a subset or superset of the other metrics? Or completely unrelated?

Reply:
I read these RPCs as neither a subset or superset but instead belonging to the same set of internal RPCs. They are simply housed in a separate file since they are auto-generated.

Reply:
I'm also confused by the relationship between this metric and `rpc_request_latency_seconds`. Building the branch to test it out.

Reply:
Think I've figured it out.

`rpc_gen.py` generates multiple *services*. All of these generated services run on the *rpc server* (plugged into the server via `rpc::simple_protocol::register_service` . These server's name are for internal communication, hence the name `internal_rpc`. There's a second server called `kafka_rpc` which is exposed to consumers.

The `rpc_request_latency_seconds` metric in `rpcgen.py` is a *per service* metric, while the `rpc_request_latency_seconds` added in `net::server` is a *per server* metric. The aggregation of the *service* metrics across the *server name* gives roughly the same result as the *server* level latency metrics. However, since both metrics have the same name, they will be part of the same metric family and be aggregated together (by the labels of whichever registered first). This is incorrect (intuitively, it should double the result).

To sum up, we *should not* add public metrics to `rpc_gen.py` in this case because we're interested in *server level* metrics (i.e. `rpc` or `kafka`).